File: src/test/java/com/github/kayjamlang/tests/expressions/array/ArrayExpressionGetTest.java
package com.github.kayjamlang.tests.expressions.array;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.GetExpression;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.VariableLinkExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class ArrayExpressionGetTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("test[2005]"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(GetExpression.class, expression.getClass());

        GetExpression arrayGet = (GetExpression) expression;
        assertSame(VariableLinkExpression.class, arrayGet.root.getClass());
        assertSame(ValueExpression.class, arrayGet.value.getClass());

        VariableLinkExpression variableLinkExpression = (VariableLinkExpression) arrayGet.root;
        assertEquals("test", variableLinkExpression.name);

        ValueExpression constant = (ValueExpression) arrayGet.value;
        assertEquals(2005, constant.value);
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/array/ArrayExpressionGetTest.scala
package com.github.kayjamlang.tests.expressions.array

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{GetExpression, ValueExpression, VariableLinkExpression}
import org.junit.Assert._
import org.junit.{BeforeClass, Test}


object ArrayExpressionGetTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer("test[2005]"))
  }
}

class ArrayExpressionGetTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = ArrayExpressionGetTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[GetExpression], expression.getClass)
    val arrayGet = expression.asInstanceOf[GetExpression]
    assertSame(classOf[VariableLinkExpression], arrayGet.root.getClass)
    assertSame(classOf[ValueExpression], arrayGet.value.getClass)
    val variableLinkExpression = arrayGet.root.asInstanceOf[VariableLinkExpression]
    assertEquals("test", variableLinkExpression.name)
    val constant = arrayGet.value.asInstanceOf[ValueExpression]
    assertEquals(2005, constant.value)
  }
}